sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java

```diff
@@ -230,14 +281,284 @@ public ReadFiles withAvroDataModel(GenericData model) {
+      }
+
+      @GetRestrictionCoder
+      public OffsetRange.Coder getRestrictionCoder() {
+        return new OffsetRange.Coder();
+      }
+
+      public long getRecordCount(
+          @Element FileIO.ReadableFile file, @Restriction OffsetRange restriction)
+          throws Exception {
+        InputFile inputFile = getInputFile(file);
+        Configuration conf = setConf();
+        ParquetReadOptions options = HadoopReadOptions.builder(conf).build();
+        ParquetFileReader reader = ParquetFileReader.open(inputFile, options);
+        long start = 0;
+        long end = 0;
+        start = restriction.getFrom();
+        end = restriction.getTo();
+        long recordCount = 0;
+        for (long i = start; i < end; i++) {
+          recordCount += reader.getRowGroups().get((int) i).getRowCount();
+        }
+        return recordCount;
+      }
+
+      public double getSize(@Element FileIO.ReadableFile file, @Restriction OffsetRange restriction)
+          throws Exception {
+        InputFile inputFile = getInputFile(file);
+        Configuration conf = setConf();
+        ParquetReadOptions options = HadoopReadOptions.builder(conf).build();
+        ParquetFileReader reader = ParquetFileReader.open(inputFile, options);
+        double size = 0;
+        for (long i = restriction.getFrom(); i < restriction.getTo(); i++) {
+          size += reader.getRowGroups().get((int) i).getTotalByteSize();
+        }
+        return size;
+      }
+    }
+
+    public static class BlockTracker extends OffsetRangeTracker {
+      private static final Logger LOG = LoggerFactory.getLogger(BlockTracker.class);
+      private long totalWork;
+      private long progress;
+      private long approximateRecordSize;
+
+      public BlockTracker(OffsetRange range, long work, long recordCount) {
+        super(range);
+        if (recordCount != 0) {
+          this.approximateRecordSize = work / recordCount;
+          this.totalWork = approximateRecordSize * recordCount;
+          this.progress = 0;
+        }
+      }
+
+      public void makeProgress() throws Exception {
+        progress += approximateRecordSize;
+        if (progress > totalWork) {
+          throw new IOException("Making progress out of range");
+        }
+      }
+
+      @Override
+      public Progress getProgress() {
+        if (this.lastAttemptedOffset == null) {
+          return Progress.from(0.0D, BigDecimal.valueOf(this.totalWork).doubleValue());
+        } else {
+          BigDecimal workRemaining =
+              BigDecimal.valueOf(this.totalWork)
+                  .subtract(BigDecimal.valueOf(this.progress), MathContext.DECIMAL128)
+                  .max(BigDecimal.ZERO);
+          BigDecimal work = BigDecimal.valueOf(this.totalWork);
+          LOG.info("total work: " + work + " work remaining: " + workRemaining);
+          return Progress.from(
+              work.subtract(workRemaining, MathContext.DECIMAL128).doubleValue(),
```

Comment:
I think we could just write `Progress.from(this.progress, this.totalWork - this.progress)`.